### PR TITLE
8327640: Allow NumberFormat strict parsing

### PR DESCRIPTION
**Pull request by justin-curtis-lu**
Please review this PR and associated [CSR](https://bugs.openjdk.org/browse/JDK-8327703) which introduces strict parsing for NumberFormat.

The concrete subclasses that will utilize this leniency value are `DecimalFormat` and `CompactNumberFormat`. Strict leniency allows for parsing to be used as an input validation technique for localized formatted numbers. The default leniency mode will remain lenient, and can only be set to strict through an intentional `setLenient(boolean)` method call. Leniency operates differently based off the values returned by `isGroupingUsed()`, `getGroupingSize()`, and `isParseIntegerOnly()`.

Below is an example of the change, the CSR can be viewed for further detail.

```
DecimalFormat fmt = (DecimalFormat) NumberFormat.getNumberInstance(Locale.US);
fmt.parse("1,,,0,,,00,.23Zabced45");  // returns 1000.23
fmt.setLenient(false);
fmt.parse("1,,,0,,,00,.23Zabced45"); // Now throws a ParseException
fmt.setGroupingSize(2);
fmt.parse("12,34,567"); // throws ParseException
fmt.setParseIntegerOnly(true)
fmt.parse("12,34.56"); // throws ParseException
fmt.parse("12,34"); // successfully returns the Number 1234
```

The associated tests are all localized, and the data is converted properly during runtime.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8327703](https://bugs.openjdk.org/browse/JDK-8327703) to be approved

### Issues
 * [JDK-8327640](https://bugs.openjdk.org/browse/JDK-8327640): Allow NumberFormat strict parsing (**Enhancement** - P4)
 * [JDK-8327703](https://bugs.openjdk.org/browse/JDK-8327703): Allow NumberFormat strict parsing (**CSR**)

### Reviewers
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**) ⚠️ Review applies to [eec992e3](https://git.openjdk.org/jdk/pull/18339/files/eec992e33642181c784fa0fc796190f69c40bb63)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18339/head:pull/18339` \
`$ git checkout pull/18339`

Update a local copy of the PR: \
`$ git checkout pull/18339` \
`$ git pull https://git.openjdk.org/jdk.git pull/18339/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18339`

View PR using the GUI difftool: \
`$ git pr show -t 18339`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18339.diff">https://git.openjdk.org/jdk/pull/18339.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18339#issuecomment-2003021355)

### CODE REVIEW

File: src/java.base/share/classes/java/text/DecimalFormat.java [hunk 1]
```diff
@@ -2371,22 +2399,34 @@ private final boolean subparse(String text, ParsePosition parsePosition,
...
 
-        // Check for suffix
+        // When strict, text should end with the suffix.
+        // When lenient, text only needs to contain the suffix.
         if (!isExponent) {
             if (gotPositive) {
-                gotPositive = text.regionMatches(position,positiveSuffix,0,
-                        positiveSuffix.length());
+                boolean containsPosSuffix =
+                        text.regionMatches(position, positiveSuffix,0, positiveSuffix.length());
```
Comment (turbanoff) [inline src/java.base/share/classes/java/text/DecimalFormat.java:2419]:
```suggestion
                        text.regionMatches(position, positiveSuffix, 0, positiveSuffix.length());
```

File: src/java.base/share/classes/java/text/DecimalFormat.java [hunk 2]
```diff
@@ -2371,22 +2399,34 @@ private final boolean subparse(String text, ParsePosition parsePosition,
...
+                boolean endsWithPosSuffix =
+                        containsPosSuffix && text.length() == position + positiveSuffix.length();
+                gotPositive = isStrict() ? endsWithPosSuffix : containsPosSuffix;
             }
             if (gotNegative) {
-                gotNegative = text.regionMatches(position,negativeSuffix,0,
-                        negativeSuffix.length());
+                boolean containsNegSuffix =
+                        text.regionMatches(position, negativeSuffix,0, negativeSuffix.length());
```
Comment (turbanoff) [inline src/java.base/share/classes/java/text/DecimalFormat.java:2426]:
```suggestion
                        text.regionMatches(position, negativeSuffix, 0, negativeSuffix.length());
```

File: src/java.base/share/classes/java/text/DecimalFormat.java [hunk 3]
```diff
@@ -2371,22 +2399,34 @@ private final boolean subparse(String text, ParsePosition parsePosition,
...
+                boolean endsWithNegSuffix =
+                        containsNegSuffix && text.length() == position  + negativeSuffix.length();
```
Comment (turbanoff) [inline src/java.base/share/classes/java/text/DecimalFormat.java:2428]:
```suggestion
                        containsNegSuffix && text.length() == position + negativeSuffix.length();
```

File: src/java.base/share/classes/java/text/DecimalFormat.java [hunk 4]
```diff
@@ -2438,7 +2479,7 @@ int subparseNumber(String text, int position,
                        boolean isExponent, boolean[] status) {
         // process digits or Inf, find decimal position
         status[STATUS_INFINITE] = false;
-        if (!isExponent && text.regionMatches(position,symbols.getInfinity(),0,
+        if (!isExponent && text.regionMatches(position, symbols.getInfinity(),0,
```
Comment (turbanoff) [inline src/java.base/share/classes/java/text/DecimalFormat.java:2482]:
```suggestion
        if (!isExponent && text.regionMatches(position, symbols.getInfinity(), 0,
```

Comment (justin-curtis-lu) [reply src/java.base/share/classes/java/text/DecimalFormat.java]: Thanks, this and the other suggestions you provided should be fixed.

File: src/java.base/share/classes/java/text/DecimalFormat.java [hunk 5]
```diff
@@ -38,6 +38,9 @@
 
 package java.text;
 
+import sun.util.locale.provider.LocaleProviderAdapter;
+import sun.util.locale.provider.ResourceBundleBasedAdapter;
+
```
Comment (naotoj) [inline src/java.base/share/classes/java/text/DecimalFormat.java:43]: Internal packages typically follow public packages.

File: src/java.base/share/classes/java/text/DecimalFormat.java [hunk 6]
```diff
@@ -2575,7 +2648,24 @@ int subparseNumber(String text, int position,
             }
         }
         return position;
+    }
+
+    // Checks to make sure grouping size is not violated. Used when strict.
```
Comment (naotoj) [inline src/java.base/share/classes/java/text/DecimalFormat.java:2653]: If it is only supposed to be used in `strict`, might be helpful to put an assertion here.

File: src/java.base/share/classes/java/text/DecimalFormat.java [hunk 7]
```diff
@@ -2575,7 +2648,24 @@ int subparseNumber(String text, int position,
...
+    private boolean isGroupingViolation(int pos, int prevGroupingPos) {
+        return isGroupingUsed() && // Only violates if using grouping
+                // Checks if a previous grouping symbol was seen.
+                prevGroupingPos != -groupingSize &&
+                // The check itself, - 1 to account for grouping/decimal symbol
+                pos - 1 != prevGroupingPos + groupingSize;
+    }
 
+    // Calculates the index that violated the grouping size
+    // Calculation is determined whether it was an under or over violation
```
Comment (naotoj) [inline src/java.base/share/classes/java/text/DecimalFormat.java:2663]: Not quite sure what `under/over violation` means here. Need more comments?

File: src/java.base/share/classes/java/text/NumberFormat.java [hunk 8]
```diff
@@ -38,8 +38,11 @@
 
 package java.text;
 
-import java.io.InvalidObjectException;
+import sun.util.locale.provider.LocaleProviderAdapter;
+import sun.util.locale.provider.LocaleServiceProviderPool;
+
```
Comment (naotoj) [inline src/java.base/share/classes/java/text/NumberFormat.java:43]: Same as above

File: src/java.base/share/classes/java/text/NumberFormat.java [hunk 9]
```diff
@@ -463,6 +476,42 @@ public void setParseIntegerOnly(boolean value) {
...
+     * UnsupportedOperationException}. Subclasses should override this method
+     * when implementing strict parsing.
+     * @throws    UnsupportedOperationException if the implementation of this
+     *            method does not support this operation
+     * @see ##leniency Leniency Section
+     * @see #setStrict(boolean)
+     * @since 23
+     */
+    public boolean isStrict() {
+        throw new UnsupportedOperationException();
```
Comment (naotoj) [inline src/java.base/share/classes/java/text/NumberFormat.java:493]: Some message might be helpful.

File: src/java.base/share/classes/java/text/NumberFormat.java [hunk 10]
```diff
@@ -463,6 +476,42 @@ public void setParseIntegerOnly(boolean value) {
...
+     * @param strict {@code true} if parsing should be done strictly;
+     *               {@code false} otherwise
+     * @throws    UnsupportedOperationException if the implementation of this
+     *            method does not support this operation
+     * @see ##leniency Leniency Section
+     * @see #isStrict()
+     * @since 23
+     */
+    public void setStrict(boolean strict) {
+        throw new UnsupportedOperationException();
```
Comment (naotoj) [inline src/java.base/share/classes/java/text/NumberFormat.java:512]: Same here

**Review by naotoj — COMMENTED** [review]
Looks good overall. Left some minor comments.
As to the tests, good to see those corner cases, but they should have unit tests for the new methods, i.e, `isStrict()` and `setStrict()`. Also I think equality/serialization for those methods should be examined.

**Comment by justin-curtis-lu** [comment]
Thank you for the review @naotoj. All of your comments should be reflected in the two previous commits.

As for the test suggestions, I added to existing tests if possible, otherwise I created new test files to address your suggestions. The new DecimalFormat equality and serialization test can definitely have other qualities tested (beyond parseStrict), but perhaps that's best for a separate issue.

File: src/java.base/share/classes/java/text/DecimalFormat.java (hunk 5 above)
Comment (justin-curtis-lu) [reply src/java.base/share/classes/java/text/DecimalFormat.java]: Fixed. (IntelliJ is adamant on having it that way, and even after I previously reverted it, it seems like it slipped through again. Oddly, it never used to do this, need to look at the settings)

**Review by naotoj — APPROVED** [review]
LGTM. Thanks for the changes.

**Pull request closed without merging**